applications/tests/function_tests/test_cuda_geo_transformer_apply_bspline_transform.cpp

```diff
@@ -0,0 +1,293 @@
+#include <gtest/gtest.h>
+#include <memory>
+#include <random>
+
+#include "reconstruction_cuda/cuda_gpu_geo_transformer.h"
+
+template< typename T >
+class GeoTransformerTest : public GeoTransformer< T > {
+public:
+
+    void applyBSplineTransformTest(int splineDegree,
+        MultidimArray<T> &output, const MultidimArray<T> &input,
+        const std::pair<Matrix1D<T>, Matrix1D<T>> &coeffs, size_t imageIdx, T outside) {
+        GeoTransformer< T >::applyBSplineTransformRef(splineDegree, output, input, coeffs, imageIdx, outside);
+    }
+};
+
+template< typename T >
+class GeoTransformerApplyBSplineTransformTest : public ::testing::Test {
+public:
+    void compare_results( double* true_values, double* approx_values, size_t size ) {
+        for ( int i = 0; i < size; ++i ) {
+            ASSERT_NEAR( true_values[i], approx_values[i], 0.000001 ) << "at index:" << i;
+        }
+    }
+
+    /*
+     * For big matrices there can be a small number of pixels whose value differs more than 0.0001
+     * from reference values, therefore I test it for lower precision with floats
+     * If tested with double values we require high precision
+    */
+    void compare_results( float* true_values, float* approx_values, size_t size ) {
+        for ( int i = 0; i < size; ++i ) {
+            ASSERT_NEAR( true_values[i], approx_values[i], 0.001f ) << "at index:" << i;
```

Comment:
For double it can be 10^-10 and it will still pass. The problem is with the precision of float, because in the kernel (both in new and old) the x (y) coordinate of pixel is converted to float and is shifted. As the coordinate gets larger, we lose precision. With precision 10^-4 there is about 600 pixels with different values in 4K picture ( 8million pixels ), but they are still in 10^-3. I wrote a bad number of zeros in our conversation. I should stick to the 10^ notation even in the code to prevent this.